drivers/flash/flash_npcx_fiu_qspi.c

```diff
@@ -0,0 +1,285 @@
+/*
+ * Copyright (c) 2023 Nuvoton Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nuvoton_npcx_fiu_qspi
+
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/flash/npcx_flash_api_ex.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/spi.h>
+#include <zephyr/dt-bindings/flash_controller/npcx_fiu_qspi.h>
+#include <soc.h>
+
+#include "flash_npcx_fiu_qspi.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(npcx_fiu_qspi, LOG_LEVEL_ERR);
+
+/* Driver convenience defines */
+#define HAL_INSTANCE(dev) \
+	((struct fiu_reg *)((const struct npcx_qspi_fiu_config *)(dev)->config)->base)
+
+/* Timeout for UMA transaction. (Unit:ms) */
+#define NPCX_UMA_TIMEOUT 10
+
+/* QSPI bus mode for Direct Read Access */
+enum npcx_qspi_fiu_read_mode {
+	 /* SPI NOR Normal read */
+	 NPCX_QSPI_FIU_READ_NOR = 0,
+	 /* SPI NOR Fast read */
+	 NPCX_QSPI_FIU_READ_FAST,
+	 /* Reserved */
+	 NPCX_QSPI_FIU_READ_RESV,
+	 /* SPI NOR Fast Read Dual I/O */
+	 NPCX_QSPI_FIU_READ_FAST_DUAL,
+};
+
+/* Device config */
+struct npcx_qspi_fiu_config {
+	/* Flash interface unit base address */
+	uintptr_t base;
+	/* Clock configuration */
+	struct npcx_clk_cfg clk_cfg;
+	/* Enable 2 external SPI devices for direct read on QSPI bus */
+	bool en_direct_access_2dev;
+};
+
+/* Device data */
+struct npcx_qspi_fiu_data {
+	/* mutex of qspi bus controller */
+	struct k_sem lock_sem;
+	/* Active device on QSPI bus */
+	const struct npcx_qspi_cfg *active;
+	/* Current Software controlled Chip-Select number */
+	int sw_cs;
+	/* Current QSPI bus operation */
+	uint32_t operation;
+};
+
+/* NPCX SPI User Mode Access (UMA) functions */
+static inline void qspi_npcx_uma_cs_level(const struct device *dev, int sw_cs, int level)
```

Comment:
should `sw_cs` and `level` be `uint8_t` and `bool`? Also wondering if it would make sense to validate `sw_cs` somehow. Maybe not necessary since it's passed from definitions in the tree but it feels risky to pass an argument straight to `BIT` and into a register, what do you think?

Reply:
That's a good idea. I will fix them later.

Reply:
Done.